### ISSUE TITLE
Support id_token_hint in OIDC and CIBA in keycloak authorization server

### ISSUE THREAD

**Issue by AxelNennker** [keycloak/keycloak#28819]
### Description

`id_token_hint` is an optional parameter in an [OpenId Connect authentication request](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) and in [Client-Initiated Backchannel Authentication](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#auth_request) 

It seems that Keycload does not support `id_token_hint`.

I found `id_token_hint` in some places in the code and other places in this repo.
But e.g. this [adoc](https://github.com/keycloak/keycloak/blob/main/docs/documentation/server_admin/topics/sso-protocols/oidc.adoc) says that only login_hint is supported.

Please support `id_token_hint`

### Motivation

Why am I asking about this? I argued in this [project´s IdentityAndConsent WG](https://github.com/camaraproject/IdentityAndConsentManagement) that id_token_hint is superior to login_hint, because the IdP can validate the signature of the id_token and therefore be sure that the sub claim was created by itself, and therefore the sub value is known and understood by the IdP. Somebody countered that keycloak does not support id_token_hint. I would like to kill that argument.

### Details

The straight forward way of implementing id_token_hint in the kc *authorization server* seems to be (happy flow):
- validate the signature of the [id token](https://openid.net/specs/openid-connect-core-1_0.html#IDToken)
- take the sub field and treat it like having received that value in the already supported login_hint parameter

Add support here?:
https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java#L61
https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java#L97

This issue is not about supporting id_token_hint in e.g. keycloak.js
https://github.com/keycloak/keycloak/blob/main/js/libs/keycloak-js/src/keycloak.js#L455
Supporting seems also be straight forward but might the resulting url might exceed length limits

```javascript
        if (options && options.id_token_hint) {
            url += '&id_token_hint=' + encodeURIComponent(options.id_token_hint);
        }
```

I am probably missing something.

**Comment by mschallar** [comment]
Wouldn't it be nice, that the "Authentication Requested User Hint" could be derived dynamically. Currently, it is only possible to select one single entry, e.g., `login_hint_token` for the realm. One could argue, that the user hint could be per keycloak client.
Or even an extension that be can be overridden/implemented and the user hint dynamically be determined?

**Comment by AxelNennker** [comment]
> Wouldn't it be nice, that the "Authentication Requested User Hint" could be derived dynamically. Currently, it is only possible to select one single entry, e.g., `login_hint_token` for the realm. One could argue, that the user hint could be per keycloak client. Or even an extension that be can be overridden/implemented and the user hint dynamically be determined?

@mschallar are you suggesting that `login_hint_token` MUST also be considered when `id_token_hint` support is added to keycloak? Otherwise creating a separate issue for `login_hint_token` seems the way forward to keep things small and focused. Do you support adding `id_token_hint` support?

**Comment by mschallar** [comment]
I suggest, that the choice of the user-hint-to-be-used should be:

1. Keycloak client specific
2. Request specific
3. Definitely not realm specific

So, imho the best idea would be a more general solution to satisfy all future needs. Currently, a realm configuration is not flexible enough.

**Comment by thomasdarimont** [comment]
@AxelNennker the id_token_hint seems to be already supported by CIBA, see: https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/grants/ciba/endpoints/BackchannelAuthenticationEndpoint.java#L270

I just gave this a quick look. The `login_hint` parameter is currently used in a lot of places, and I guess some of those places would need to be changed to also to check the "id_token_hint" additionally. 

For that I like @mschallar suggestion to have a generic solution that "resolves a username / sub" from hints in the request.

**Comment by AxelNennker** [comment]
I agree with that having something "resolves a username / sub" from hints in the request should be done.

One reason for that is, that OIDC and CIBA allow exactly _one_ hint to be in the request. Requests with more than one hint are an client error.

**Comment by thomasdarimont** [comment]
@AxelNennker I think it makes sense that the requests with more than one hint should result in an client error as it forms an ambiguity, however do you know if the spec mandates that behaviour? 

I couldn't find this in the https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest document.

**Comment by AxelNennker** [comment]
OIDC spec does not seem to specify how the OP should handle both id_token_hint and login_hint being parameters in the request.

I suggest the following:
```
If `prompt=none` and there are both an id_token_hint and a login_hint parameter in the request, than the id_token_hint takes precedent and the login_hint parameter SHOULD be ignored. The authorization server MUST not try to use the id_token followed by trying the login_hint or vice versa.
If there is no prompt parameter or its value is other then `none` and there are both an id_token_hint and a login_hint parameter in the request, then it is RECOMMENDED that the authorization server uses the id_token_hint.
```

Hints are _only_ hints and if there is some user interaction (prompt not none), then the OP could ignore both hints and let the user enter their identifier. I think picking the id_token is better. 
And the id_token MUST not be ignored. Meaning if there is an id_token the signature MUST be valid and the aud must identify the OP. If the id_token signature is invalid or the aud does not identify the OP, then invalid_request MUST be returned.

**Comment by thomasdarimont** [comment]
I like that addition and I agree with most of your suggestions, but I think that's a question that should also be discussed on the openid WG mailinglist.
Regarding:
> ...Meaning if there is an id_token the signature MUST be valid and the aud must identify the OP...
https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest for id_token_hint mentions:
` The Authorization Server need not be listed as an audience of the ID Token when it is used as an id_token_hint value.`
Which seems to not require a hard aud claim check here. 

However, it might make sense to check whether the ID token issuer is the OP.

**Comment by AxelNennker** [comment]
> Yes, one and one only PER request, not PER keycloak client.

In this issue I would like to focus on the server side.

**Comment by AxelNennker** [comment]
> I like that addition and I agree with most of your suggestions, but I also think that's a question that should also be discussed on the openid WG mailinglist. Regarding:
> 
> > ...Meaning if there is an id_token the signature MUST be valid and the aud must identify the OP...
> 
> OIDC docs https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest for id_token_hint mentions: ` The Authorization Server need not be listed as an audience of the ID Token when it is used as an id_token_hint value.` Which seems to not require a hard aud claim check here.
> 
> However, it might make sense to check whether the ID token issuer is the OP.

Please chime in here: https://lists.openid.net/pipermail/openid-specs-ab/2024-April/010298.html

**Comment by AxelNennker** [comment]
I guess that `aud` is not really worth checking if the signature is valid and the signer is the OP.

**Comment by AxelNennker** [comment]
I retract " and the aud must identify the OP".

**Comment by mschallar** [comment]
> OIDC spec does not seem to specify how the OP should handle both id_token_hint and login_hint being parameters in the request.
> 
> I suggest the following:
> 
> ```
> If `prompt=none` and there are both an id_token_hint and a login_hint parameter in the request, than the id_token_hint takes precedent and the login_hint parameter SHOULD be ignored. The authorization server MUST not try to use the id_token followed by trying the login_hint or vice versa.
> If there is no prompt parameter or its value is other then `none` and there are both an id_token_hint and a login_hint parameter in the request, then it is RECOMMENDED that the authorization server uses the id_token_hint.
> ```
> 
> Hints are _only_ hints and if there is some user interaction (prompt not none), then the OP could ignore both hints and let the user enter their identifier. I think picking the id_token is better. And the id_token MUST not be ignored. Meaning if there is an id_token the signature MUST be valid and the aud must identify the OP. If the id_token signature is invalid or the aud does not identify the OP, then invalid_request MUST be returned.

Just to rule out any misunderstanding: Are you talking about a CIBA use case here, or a general OIDC one?
Because:

> CIBA says:
> 
> > Because in the CIBA flow, the OP does not have an interaction with the end-user through the consumption device, it is REQUIRED that the Client provides one (**and only one**) of the hints specified above in the authentication request, that is "login_hint_token", "id_token_hint" or "login_hint".

I think this snipped from the CIBA Spec is pretty clear, on how to handle requests containing more then one of the listed parameters for CIBA.

**Comment by AxelNennker** [comment]
I think keycloak server should support id_token_hint in both CIBA and OIDC.

This could be implemented in two steps: CIBA first - with an eye on the future OIDC implementation - and OIDC support later. There is some common code probably.

**Comment by AxelNennker** [comment]
side note: When I try building keycloak on my MacBook it fails with strange error messages
`HostnameV2Test.testAdminLocal:151 » UnknownHost 127.0.0.1.nip.io: nodename nor servname provided, or not known`
It is a little discouraging when a developer, me, clones the repo, does a `./mvnw clean install` and it fails.

**Comment by thomasdarimont** [comment]
nip.io is blocked by some dns configurations. You could add an entry like `127.0.0.1 localhost 127.0.0.1.nip.io` to your `/etc/hosts` file. I think this hint could be added to the developer guide.

**Comment by AxelNennker** [comment]
Regarding id_token_hint support in CIBA in resolveUser
https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/grants/ciba/endpoints/BackchannelAuthenticationEndpoint.java#L270

I think that implementation does not follow the CIBA standard because the implementation allows multiple hints and give precedent to login_hint. Also the signature of the id_token is not validated.
Maybe the CIBA only-one-hint rule is enforced somewhere else?

Regarding 
> Currently you have to provide your own: > org.keycloak.protocol.oidc.grants.ciba.resolvers.CIBALoginUserResolver#getUserFromIdTokenHint implementation.

I think most keycloak admins/customer do not think about re-used login identifiers. There are some pitfalls there that might enable fraud. `sub` is locally unique and should never be re-assigned. But I do not see the required id token field `sub` in the org.keycloak.representations.IDToken class. Huh?

**Comment by randomstuff** [comment]
> Why am I asking about this? I argued in this [project´s IdentityAndConsent WG](https://github.com/camaraproject/IdentityAndConsentManagement) that <mark>id_token_hint is superior to login_hint</mark>, because the IdP can validate the signature of the id_token and therefore be sure that the sub claim was created by itself, and therefore the sub value is known and understood by the IdP. Somebody countered that keycloak does not support id_token_hint. I would like to kill that argument.

For what it's worth, I am not absolutely fan of the idea of passing the ID token around through the front-channel (in the OIDC authentication request) and it's probably better that way.

When using the code exchange flow (which is the recommended flow nowadays), the ID token is actually not exposed to the user-agent front-end.

On the other hand, some usages allow trading an ID token in exchange for other tokens:

* In the case of Keycloak, the ID token [can be exchanged for a UMA RPT token](http://openid.net/specs/openid-connect-core-1_0.html#IDToken), with a `POST …/token` with `grant_type=urn:ietf:params:oauth:grant-type:uma-ticket&ticket={TICKET}&claim_token_format=http://openid.net/specs/openid-connect-core-1_0.html#IDToken&claim_token={ID_TOKEN}` (without any client authentication). → An exfltrated  ID token could be used by a malicious actor in order to obtain UMA 2.0 RPT.

* [Token Exchange](https://datatracker.ietf.org/doc/html/rfc8693) could be used to exchange an ID token against other tokens. → An exfiltrated could have other unexpected impacts.

Therefore, I would argue that exposing the ID token to the front-end by sending it through the front-channel is not worth the risk.

One important benefit of passing the ID token around, is that it makes it possible to reference the user by subject instead of referencing the user by login.

**Comment by AxelNennker** [comment]
Yes, id_token_hint is about referencing the user by subject. The AZ created and signed the id_token. So the AZ can validate that signature and be sure that the data inside the id_token, including sub, is unmodified (and was created by itself).

OIDC states this regarding "id_token_hint":
"...If the ID Token received by the RP from the OP is encrypted, to use it as an id_token_hint, the Client MUST decrypt the signed ID Token contained within the encrypted ID Token. The Client MAY re-encrypt the signed ID token to the Authentication Server using a key that enables the server to decrypt the ID Token and use the re-encrypted ID token as the id_token_hint value...."

Keycloak could require that the id_token_hint is encrypted to itself if the front-channel is used in OIDC authorization code flow if you feel the id_token data must be protected.
In CIBA the request is server-to-server anyway and does not go through the front-channel, but, I guess, re-encrypting does not hurt if it is required for auth code flow anyway.

So, when it comes to keycloak support for id_token_hint: 
- CIBA does sent the id_token_hint through the front-channel and security and privacy concerns stated by @randomstuff do not apply, right? 
- For keycloak's support for id_token_hint in OIDC auth code flow keycloak could require encryption of if_token

**Comment by randomstuff** [comment]
> CIBA does sent the id_token_hint through the front-channel and security and privacy concerns stated by @randomstuff do not apply, right?

Oh, yes you are right! :+1:

**Comment by embesozzi** [comment]
Hi guys,
I need the feature for using the login_hint [1] in the authorize endpoint.
Is this feature available? If not, I’m going to implement it in the custom authenticator to pick this up.

[1] https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest

**Comment by oculos** [comment]
This is extremely useful now that Microsoft allows third party idp's for MFA: 
https://learn.microsoft.com/en-us/entra/identity/authentication/concept-authentication-external-method-provider#configure-a-new-external-authentication-provider-with-microsoft-entra-id

This means that if you federate Entra to Keycloak, you still might be dependent on Entra's MFA. But if id_token_hint is supported, you can reuse Keycloak's own two factor.

**Comment by embesozzi** [comment]
I ended up implementing a custom authenticator to integrate MS Entra ID.
